### ISSUE TITLE
Implement JSON serializers for all assets

### ISSUE THREAD

**Issue by goblinJoel** [rpgtoolkit/editor#18]
Assets should be able to save/load in JSON format to provide a flexible format that is human-readable and easy to change.

**Comment by goblinJoel** [comment]
An abstract JSON serializer has been made. Individual ones (for example, an Enemy JSON serializer) can extend it without having to worry about much of the boilerplate.

**Comment by goblinJoel** [comment]
AssetHandle now has getReader() and getWriter() for string files (previously only handled byte streams). This allows us to read/write JSON using AssetHandles. This is good.

**Comment by goblinJoel** [comment]
I've been using SpecialMove as a test case. It now supports JSON serialization/deserialization, but binary has regressed for the moment.

**Comment by chutchinson** [comment]
There are serious problems with using simple Reflection-based serialization, please read my post on the official forums: http://rpgtoolkit.net/community/discussion/46834/25/

**Comment by goblinJoel** [comment]
This was answered in the forums, but I'm not doing reflection. I think Euix and I are on the same page with JSON, or close to it. Assets need to specify which JSON entries get written and parsed. The abstract class just handles the stuff for the file i/o and things like that.

**Comment by swordmaster2k** [comment]
This needs to be broken down into multiple taks/issues so that the current progress can be tracked in the milestone.

**Comment by goblinJoel** [comment]
@swordmaster2k We should see about getting it to fill in the file extension automatically. Had trouble testing the enemy editor because I didn't realize I had to type in the .ene.json extension manually.

**Comment by goblinJoel** [comment]
@swordmaster2k We should consider how we want to handle missing values. Right now we're using a lot of optFoo(), but we're not handling the cases where it actually does come back with null, NaN, etc. For example, when we pass the result of optJSONArray() into getStringArrayList(), it doesn't handle null.

**Comment by swordmaster2k** [comment]
Yes I agree, editor should default to the .json extensions when saving.
